### PR TITLE
The Future

### PR DESCRIPTION
**Pull request by madcowswe**
### Added
* Automated test script `run_tests.py`
* Encoder can now go forever in velocity/torque mode due to using circular encoder space.
* Protocol supports function return values
### Changed
* GPIO3 not sensitive to edges by default

### CODE REVIEW

**Comment by madcowswe** [comment]
@samuelsadok If you have a good way to implement the function return values in not as hacky way, then that would be cool!

File: Firmware/Board/v3/Src/stm32f4xx_it.c [hunk 1]
```diff
@@ -341,6 +341,14 @@ void EXTI4_IRQHandler(void)
   HAL_GPIO_EXTI_IRQHandler(GPIO_PIN_4);
 }
 
+/**
+* @brief This function handles EXTI lines 5-9 interrupt.
+*/
+void EXTI9_5_IRQHandler(void)
+{
+  HAL_GPIO_EXTI_IRQHandler(GPIO_PIN_9);
```
Comment (samuelsadok) [inline Firmware/Board/v3/Src/stm32f4xx_it.c]: what if one of pins 5-8 triggered the interrupt?

File: Firmware/Makefile [hunk 2]
```diff
@@ -5,7 +5,7 @@
 BUILD_DIR = build
 FIRMWARE = $(BUILD_DIR)/ODriveFirmware.elf
 FIRMWARE_HEX = $(BUILD_DIR)/ODriveFirmware.hex
-PROGRAMMER_HEX := $(shell echo $(PROGRAMMER) | sed -e 's/.\{2\}/\\x&/g')
+PROGRAMMER_HEX := $(shell echo $(PROGRAMMER) | sed -e 's/.\\{2\\}/\\\\x&/g')
```
Comment (samuelsadok) [inline Firmware/Makefile]: This breaks it on my machine
```
…ects/ODrive/Firmware (sam_oskar_testing) $ make flash PROGRAMMER=483f6d06483f53573328103f
 16) [0.026s] arm-none-eabi-size build/ODriveFirmware.elf                                                                                                    
   text	   data	    bss	    dec	    hex	filename
 128036	 138728	  26080	 292844	  477ec	build/ODriveFirmware.elf
 [     ETA=<1s     ] 100%                                                                                                                                    
openocd -f interface/stlink-v2.cfg -c 'hla_serial 483f6d06483f53573328103f' -f target/stm32f4x.cfg -c init \                                                 
	-c 'reset halt' \
	-c 'flash write_image erase build/ODriveFirmware.elf' \
	-c 'reset run' \
	-c exit
[...]
```

File: Firmware/MotorControl/encoder.hpp [hunk 3]
```diff
@@ -50,10 +51,12 @@ class Encoder {
...
     bool index_found_ = false;
     bool is_ready_ = false;
-    int32_t state_ = 0;
+    int32_t shadow_count_ = 0;
+    int32_t count_in_cpr_ = 0;
     int32_t offset_ = 0;
     float phase_ = 0.0f;    // [rad]
-    float pll_pos_ = 0.0f;  // [rad]
+    float pos_estimate_ = 0.0f;  // [rad]
+    float pos_cpr = 0.0f;  // [rad]
```
Comment (samuelsadok) [inline Firmware/MotorControl/encoder.hpp]: this should probably be `pos_cpr_` for consistency

File: Firmware/MotorControl/motor.cpp [hunk 4]
```diff
@@ -108,9 +116,9 @@ bool Motor::check_DRV_fault() {
         // Update DRV Fault Code
         drv_fault_ = DRV8301_getFaultType(&gate_driver_);
         // Update/Cache all SPI device registers
-        DRV_SPI_8301_Vars_t* local_regs = &gate_driver_regs_;
-        local_regs->RcvCmd = true;
-        DRV8301_readData(&gate_driver_, local_regs);
+        // DRV_SPI_8301_Vars_t* local_regs = &gate_driver_regs_;
```
Comment (samuelsadok) [inline Firmware/MotorControl/motor.cpp]: was this commented out just for testing? (same in protocol exports)

**Comment by samuelsadok** [comment]
@madcowswe yep, the function return values are now implemented more cleanly

File: Firmware/MotorControl/motor.cpp (hunk 4 above)
Comment (samuelsadok) [reply Firmware/MotorControl/motor.cpp]: as per discussion, this is on purpose to make the fault register fetch as fast as possible

**Pull request merged**